kinesis/src/main/scala/akka/stream/alpakka/kinesis/ShardSettings.scala

```diff
@@ -4,93 +4,121 @@
 
 package akka.stream.alpakka.kinesis
 
-import java.util.Date
-import java.util.concurrent.TimeUnit
-
 import com.amazonaws.services.kinesis.model.ShardIteratorType
 
-import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.duration._
+import akka.util.JavaDurationConverters._
+
+final class ShardSettings private (
+    val streamName: String,
+    val shardId: String,
+    val shardIteratorType: com.amazonaws.services.kinesis.model.ShardIteratorType,
+    val startingSequenceNumber: Option[String],
+    val atTimestamp: Option[java.time.Instant],
+    val refreshInterval: scala.concurrent.duration.FiniteDuration,
+    val limit: Int
+) {
+  require(
+    limit >= 1 && limit <= 10000,
+    "Limit must be between 0 and 10000. See: http://docs.aws.amazon.com/kinesis/latest/APIReference/API_GetRecords.html"
+  )
+  shardIteratorType match {
+    case ShardIteratorType.AFTER_SEQUENCE_NUMBER | ShardIteratorType.AT_SEQUENCE_NUMBER =>
+      require(
+        startingSequenceNumber.nonEmpty,
+        "a starting sequence number must be set (try using just `withStartingSequenceNumber` or `withStartingAfterSequenceNumber`)"
+      )
+    case ShardIteratorType.AT_TIMESTAMP =>
+      require(atTimestamp.nonEmpty, "a timestamp must be set (try using just `withAtTimestamp`)")
+    case _ => ()
+  }
 
-object ShardSettings {
+  def withStreamName(value: String): ShardSettings = copy(streamName = value)
+  def withShardId(value: String): ShardSettings = copy(shardId = value)
+  def withShardIteratorType(value: ShardIteratorType): ShardSettings = copy(shardIteratorType = value)
 
   /**
-   * Java API: Create settings using the default configuration
+   * Sets `shardIteratorType` to `AT_SEQUENCE_NUMBER` and uses the given value as starting sequence number.
    */
-  def create(streamName: String, shardId: String): ShardSettings =
-    ShardSettings(streamName,
-                  shardId,
-                  ShardIteratorType.LATEST,
-                  None,
-                  None,
-                  FiniteDuration.apply(1L, TimeUnit.SECONDS),
-                  500);
+  def withStartingSequenceNumber(value: String): ShardSettings =
+    copy(shardIteratorType = ShardIteratorType.AT_SEQUENCE_NUMBER, startingSequenceNumber = Option(value))
 
   /**
-   * Java API: Create settings using the default configuration
+   * Sets `shardIteratorType` to `AFTER_SEQUENCE_NUMBER` and uses the given value as starting sequence number.
    */
-  def create(streamName: String,
-             shardId: String,
-             shardIteratorType: ShardIteratorType,
-             refreshInterval: FiniteDuration,
-             limit: Integer) =
-    ShardSettings(streamName, shardId, shardIteratorType, None, None, refreshInterval, limit)
+  def withStartingAfterSequenceNumber(value: String): ShardSettings =
+    copy(shardIteratorType = ShardIteratorType.AFTER_SEQUENCE_NUMBER, startingSequenceNumber = Option(value))
```

Comment:
These are really useful when the type and some property needs to be set in coordination.